crates/sui-graphql-client/src/faucet.rs

```diff
@@ -0,0 +1,235 @@
+            StatusCode::SERVICE_UNAVAILABLE => {
+                bail!("Faucet service is currently overloaded or unavailable. Please try again later.");
+            }
+            status_code => {
+                bail!("Faucet request was unsuccessful: {status_code}");
+            }
+        }
+    }
+
+    /// Request gas from the faucet and wait until the request is completed and token is
+    /// transferred. Returns `FaucetReceipt` if the request is successful, which contains the list
+    /// of tokens transferred, and the transaction digest.
+    pub async fn request_and_wait(
+        &self,
+        address: Address,
+    ) -> Result<Option<FaucetReceipt>, anyhow::Error> {
+        let request_id = self.request(address).await?;
+        if let Some(request_id) = request_id {
+            let status_url = Url::parse(
+                self.faucet_url
+                    .to_string()
+                    .replace("gas", "status")
+                    .as_str(),
+            )
+            .expect("Invalid faucet status URL");
+
+            let poll_response = tokio::time::timeout(FAUCET_REQUEST_TIMEOUT, async {
+                let mut interval = tokio::time::interval(FAUCET_POLL_INTERVAL);
+                loop {
+                    interval.tick().await;
+                    eprint!("Polling faucet status...");
+                    let req = self
+                        .request_status(request_id.clone(), status_url.clone())
+                        .await;
+
+                    if let Ok(Some(poll_response)) = req {
+                        eprintln!("response: {:?}", poll_response);
+                        match poll_response.status {
+                            BatchSendStatusType::Succeeded => {
+                                break Ok(poll_response);
+                            }
+                            BatchSendStatusType::Discarded => {
+                                break Ok(BatchSendStatus {
+                                    status: BatchSendStatusType::Discarded,
+                                    transferred_gas_objects: None,
+                                });
+                            }
+                            BatchSendStatusType::Inprogress => {
+                                continue;
+                            }
+                        }
+                    } else if req.is_err() {
+                        break Err(anyhow::anyhow!("Faucet request failed. {:?}", req.err()));
+                    }
+                }
+            })
+            .await
+            .map_err(|_| anyhow::anyhow!("Faucet request timed out"))??;
+            Ok(poll_response.transferred_gas_objects)
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Request gas from a custom faucet service.
+    pub async fn request_url(
+        &self,
+        address: Address,
+        url: Url,
+    ) -> Result<Option<String>, anyhow::Error> {
+        self.request_impl(address, &url).await
+    }
```

Comment:
We probably don't need this if people can construct a client that is bespoke to the url they care about